### ISSUE TITLE
Erroneous "missing super() call in constructor" error thrown during transformation

### ISSUE THREAD

**Issue by TimothyGu** [babel/babel#6370]
This is a bug report of transform-es2015-classes.

### Input Code

```js
// Any parent class would work.
class Child extends Set {
  constructor() {}
}
```

([REPL link](https://babeljs.io/repl/#?babili=false&browsers=&build=&builtIns=false&code_lz=MYGwhgzhAEDCAWBLEATaBTAHgF3QOxRgGV1toBvAKGmmAHs8JsAnAV2GzuYAoBKCgL6UBQA&debug=false&circleciRepo=&evaluate=false&lineWrap=false&presets=es2015&targets=&version=6.26.0))

or 

```js
class Child extends Set {
  constructor() {
    return {};
  }
}
```

([REPL link](https://babeljs.io/repl/#?babili=false&browsers=&build=&builtIns=false&code_lz=MYGwhgzhAEDCAWBLEATaBTAHgF3QOxRgGV1toBvAKGmmAHs8JsAnAV2GzuYAoBKC6jWjNSrZngoBfANyDJlSUA&debug=false&circleciRepo=&evaluate=false&lineWrap=false&presets=es2015&targets=&version=6.26.0))

### Babel Configuration (.babelrc, package.json, cli command)

"es2015" preset enabled in REPL.

### Expected Behavior
Both code fragments should parse without error. At runtime, the class in the first code fragment, when being constructed, should throw a ReferenceError, and the class in the second code fragment, when being constructed, should return `{}` and not throw any error.

From the spec, for good measure. It's self-evident that all of these are runtime semantics, not static semantics or early errors, the latter of which seems to be what Babel is implementing.

- `new Child()` -> [Evaluation of *NewExpression*](https://tc39.github.io/ecma262/#sec-new-operator-runtime-semantics-evaluation)

   - [EvaluateNew](https://tc39.github.io/ecma262/#sec-evaluatenew) step 8

      - [Construct](https://tc39.github.io/ecma262/#sec-construct)(`Child`, « ») step 5

         - [ECMAScript Function Object.[[Construct]]](https://tc39.github.io/ecma262/#sec-ecmascript-function-objects-construct-argumentslist-newtarget)(`Child`, « ») step 6

            - [PrepareForOrdinaryCall](https://tc39.github.io/ecma262/#sec-prepareforordinarycall)(`Child`, `Child`) step 8

               - [NewFunctionEnvironment](https://tc39.github.io/ecma262/#sec-newfunctionenvironment)(`Child`, `Child`) step 7

                  > Else, set *envRec*.[[ThisBindingStatus]] to **`"uninitialized"`**.

         - [ECMAScript Function Object.[[Construct]]](https://tc39.github.io/ecma262/#sec-ecmascript-function-objects-construct-argumentslist-newtarget)(`Child`, « ») steps 7-12

           During these steps, `super()` is not called, which is the only way to invoke [Function Environment Record.BindThisValue](https://tc39.github.io/ecma262/#sec-bindthisvalue)(*V*), which sets *envRec*.[[ThisBindingStatus]] to **`"initialized"`**.

         - For the second test case only: [ECMAScript Function Object.[[Construct]]](https://tc39.github.io/ecma262/#sec-ecmascript-function-objects-construct-argumentslist-newtarget)(`Child`, « ») step 13

            > If *result*.[[Type]] is return, then
            >
            > 1. If Type(*result*.[[Value]]) is Object, return NormalCompletion(*result*.[[Value]]).

         - For the first test case only: [ECMAScript Function Object.[[Construct]]](https://tc39.github.io/ecma262/#sec-ecmascript-function-objects-construct-argumentslist-newtarget)(`Child`, « ») step 15

            - [Function Environment Record.GetThisBinding](https://tc39.github.io/ecma262/#sec-function-environment-records-getthisbinding)() step 3:
   
               > If *envRec*.[[ThisBindingStatus]] is **`"uninitialized"`**, throw a **ReferenceError** exception.

### Current Behavior
Babel throws the following error during transformation:

```
repl: missing super() call in constructor
  1 | class Child extends Parent {
> 2 |   constructor() {}
    |   ^
  3 | }
```

### Possible Solution

Transform the first test case to something like

```js
function Child() {
  _classCallCheck(this, Child);
  throw new ReferenceError("super() hasn't been called");
}
```

Transform the second test case to something like

```js
function Child() {
  _classCallCheck(this, Child);
  return {};
  // Optionally, if it's too much work to implement static analysis for `return` statements
  // throw new ReferenceError("super() hasn't been called");
}
```

### Context

https://github.com/tmpvar/jsdom/issues/2007

### Your Environment

| software         | version(s)
| ---------------- | -------
| Babel            |  v6.26.0 (in the REPL)
| browser             | 62.0.3202.18 (Official Build) beta (64-bit)
| Operating System   | Linux 4.9.0-3-amd64 SMP Debian 4.9.30-2+deb9u5 (2017-09-19) x86_64 GNU/Linux

**Comment by hzoo** [comment]
Thanks for the report @TimothyGu! You can test master with `/repl/build/master`

Looks like the [2nd case works fine in 7.x/master](https://babeljs.io/repl/build/master#?babili=false&browsers=&build=&builtIns=false&code_lz=MYGwhgzhAEDCAWBLEATaBTAHgF3QOxRgGV1toBvAKGmmAHs8JsAnAV2GzuYAoBKC6jWjNSrZngoBfANyDJlSUA&debug=false&circleciRepo=&evaluate=false&lineWrap=false&presets=es2015&targets=&version=7.0.0-beta.2)

but the first case still fails

I thought @jridgewell made a fix for this earlier, and it looks like that was to handle the 2nd case https://github.com/babel/babel/pull/6189

**Comment by TimothyGu** [comment]
So this is how the class looks like under 7.x

```js
class Child extends Set {
  constructor() {
    if (true) return {};
  }
}
```

```js
var Child =
/*#__PURE__*/
function (_Set) {
  _inherits(Child, _Set);

  function Child() {
    var _this;

    _classCallCheck(this, Child);

    if (true) return _possibleConstructorReturn(_this, {});
    return _possibleConstructorReturn(_this);
  }

  return Child;
}(Set);
```

So it looks like all the necessary pieces are there to fix the first issue.

----

with 

```js
class Child extends Set {
  constructor() {}
}
```

the transpiled class can look like

```js
var Child =
/*#__PURE__*/
function (_Set) {
  _inherits(Child, _Set);

  function Child() {
    var _this;

    _classCallCheck(this, Child);

    return _possibleConstructorReturn(_this);
  }

  return Child;
}(Set);
```

In other words, I think simply removing 

https://github.com/babel/babel/blob/91cde0148d16afe9f9192953084f0b668885cdf4/packages/babel-plugin-transform-es2015-classes/src/vanilla.js#L451-L453

should fix this problem.

**Comment by jridgewell** [comment]
The first is absolutely an error.

The second is fixed in master.

**Comment by nicolo-ribaudo** [comment]
@jridgewell I think the first should be a runtime error, while currently it is reported as a static error.

**Comment by TimothyGu** [comment]
In fact, even the following should be treated as a runtime error:

```js
class Child extends Set {
  constructor() {
    this.a = 123
  }
}
```

It follows the same logic as what I wrote up in https://github.com/babel/babel/issues/6370#issue-262275885, and parses successfully with no early errors in all JavaScript engines I can test besides Babel.

**Comment by piranna** [comment]
Any update on this? I'm facing the same problem.

**Comment by nicolo-ribaudo** [comment]
Fixed by #6467 (in Babel 7).
